src/ixdat/tools.py

```diff
@@ -71,3 +80,159 @@ def list_is_close(list_one, list_two):
+
+    Args:
+        last_supported_release (str): The name of the last version when the deprecated
+            functionality was fully supported e.g. "1.2.3"
+        update_message (str): A message to the user with instructions on how to upgrade
+            to avoid the deprecated functionality
+
+    Keyword Args:
+        hard_deprecation_release (str): The release with which the deprecation will raise
+            exceptions rather than issue warnings e.g. "1.4.0". The default None means
+            that this deprecation will remain soft for the foreseeable future.
+        remove_release (str): The release with which the deprecated functionality is
+            planned for deletion
+        kwarg_name (str): If given, is the name of the keyword argument which is
+            deprecated. If not given it is assumed that the entire callable is
+            deprecated.
+
+    Examples:
+
+    Used to deprecate a class::
+     @deprecate("1.2.3", "Please use `MyNewClass` instead", "1.4.0", "2.0.0")
+     class MyClass:
+         ...
+
+    Used to deprecate a method or an argument in a method::
+        class MyClass:
+
+            @deprecate("1.2.3", "Please use `mynewmethod` instead", "1.4.0", "2.0.0")
+            def mymethod(self):
+                ...
+
+            @classmethod
+            @deprecate(
+                "1.2.3",
+                "Please use `kwargb` instead",
+                "1.4.0",
+                "2.0.0",
+                kwarg_name="kwarga"
+            )
+            def mymethod(cls, arg, kwarga=None, kwargb=None):
+                ...
+
+    .. note::
+       In the example above, when this decorator is applied to a class method or
+       static method, it must be applied as the first decorator (closest to the def).
+
+    """
+
+    def decorator(callable_):
+        """Decorate a callable with"""
+        # Form an identity string for the object which is being decorated, which is used
+        # in the message to the user
+        if inspect.isclass(callable_):
+            identity = f"class '{callable_.__qualname__}'"
+        else:
+            if "." in callable_.__qualname__:
+                identity = f"method '{callable_.__qualname__}'"
+            else:
+                identity = f"function '{callable_.__qualname__}'"
+
+        compound_message = _construct_deprecation_message(
+            identity,
+            last_supported_release,
+            update_message,
+            hard_deprecation_release,
+            remove_release,
+            kwarg_name,
+        )
+
+        # Get the argument signature of the callable
+        callable_signature = inspect.signature(callable_)
+
+        @wraps(callable_)
+        def inner_function(*args, **kwargs):
+            # Form bound arguments, so both args and kwargs gets mapped to their name
+            bound_arguments = {}
+            if kwarg_name:
+                bound_arguments = callable_signature.bind(*args, **kwargs).arguments
```

Comment:
Do I understand right that this binding `*args` and `**kwargs` to the callable's signature is there to enable deprecation of a positional argument? If we were only deprecating use of `**kwargs`, then it would siffice for the condition below to just be `if kwarg_name in kwargs`, right?
I can't think of a situation where a positional argument would be deprecated without deprecating the function, in which case this might be overkill. But such a situation will probably come and we'll be happy for it :)

Reply:
Ah, there actually is a good reason for this. It's one of those cases where Python dynamic nature is ... challanging. I've actually tried and failed to solve this problem before. The problem is that you may reach a kwarg as an arg i.e if we want to deprecate `c` in the following function:
```python
def myfunc(a, b, c=None):
    pass

myfunc(1, 2, 3)
```
is allowed and will provide a value for `c` as an arg, no a kwarg. Therefore this trick is needed.

Reply:
This particular problem is of course prevented by the use of kwarg_only arguments, which you have developed the excellent habit of using to great effect, but I thought it was better if the decorator covered the other use case as well.